Uninformative error message when casting usize to pointer to generic `?Sized` type

Trying to compile this code
```rust
pub fn bogus_ptr<U: ?Sized>() -> *const U {
    0usize as *const U
}
```
produces this error message
```
error[E0606]: casting `usize` as `*const U` is invalid
  --> <source>:2:5
  |
2 |     0usize as *const U
  |     ^^^^^^^^^^^^^^^^^^
```
It would be better if the error message stated that `usize` cannot be cast to a `T: ?Sized` type.

### Meta

Tried the code in these Rust versions and they all produced the same error message:

- 1.37.0 stable

- 1.45.0 stable

- 1.46.0-beta.1 (2020-07-15 e51b714db8ff82ac38ea)

- 1.47.0-nightly (2020-07-22 bbebe7351fcd29af1eb9)

Current output:

```
error[E0606]: cannot cast `usize` to a pointer that may be wide
 --> src/lib.rs:4:15
  |
4 |     0usize as *const U
  |               ^^^^^^^^ creating a `*const U` requires both an address and type-specific metadata
```

--- a/compiler/rustc_hir_typeck/src/cast.rs
+++ b/compiler/rustc_hir_typeck/src/cast.rs
@@ -572,6 +572,17 @@ impl<'a, 'tcx> CastCheck<'tcx> {
                 let metadata = known_metadata.unwrap_or("type-specific metadata");
                 let known_wide = known_metadata.is_some();
                 let span = self.cast_span;
+                let param_note = (!known_wide)
+                    .then(|| match cast_ty.kind() {
+                        ty::RawPtr(pointee, _) => match pointee.kind() {
+                            ty::Param(param) => {
+                                Some(errors::IntToWideParamNote { param: param.name })
+                            }
+                            _ => None,
+                        },
+                        _ => None,
+                    })
+                    .flatten();
                 fcx.dcx().emit_err(errors::IntToWide {
                     span,
                     metadata,
@@ -579,6 +590,7 @@ impl<'a, 'tcx> CastCheck<'tcx> {
                     cast_ty,
                     expr_if_nightly,
                     known_wide,
+                    param_note,
                 });
             }
             CastError::UnknownCastPtrKind | CastError::UnknownExprPtrKind => {

--- a/compiler/rustc_hir_typeck/src/errors.rs
+++ b/compiler/rustc_hir_typeck/src/errors.rs
@@ -590,6 +590,14 @@ pub(crate) struct IntToWide<'tcx> {
     )]
     pub expr_if_nightly: Option<Span>,
     pub known_wide: bool,
+    #[subdiagnostic]
+    pub param_note: Option<IntToWideParamNote>,
+}
+
+#[derive(Subdiagnostic)]
+#[note("the type parameter `{$param}` is not known to be `Sized`, so this pointer may be wide")]
+pub(crate) struct IntToWideParamNote {
+    pub param: Symbol,
 }
 
 #[derive(Subdiagnostic)]

--- a/tests/ui/cast/fat-ptr-cast.stderr
+++ b/tests/ui/cast/fat-ptr-cast.stderr
@@ -81,6 +81,8 @@ LL |     let s = 0 as *const T;
    |             -    ^^^^^^^^ creating a `*const T` requires both an address and type-specific metadata
    |             |
    |             consider casting this expression to `*const ()`, then using `core::ptr::from_raw_parts`
+   |
+   = note: the type parameter `T` is not known to be `Sized`, so this pointer may be wide
 
 error: aborting due to 11 previous errors
 

